File: src/main/kotlin/com/airsoft/tactic/controller/AuthController.kt
package com.airsoft.tactic.controller

import com.airsoft.tactic.dto.request.AppleAuthRequest
import com.airsoft.tactic.dto.request.GoogleAuthRequest
import com.airsoft.tactic.dto.request.LogoutRequest
import com.airsoft.tactic.dto.request.RefreshTokenRequest
import com.airsoft.tactic.dto.response.AuthResponse
import com.airsoft.tactic.exception.ApiResponse
import com.airsoft.tactic.service.AuthService
import jakarta.validation.Valid
import org.springframework.http.ResponseEntity
import org.springframework.security.core.Authentication
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController
import java.util.UUID

@RestController
@RequestMapping("/v1/auth")
class AuthController(private val authService: AuthService) {

    @PostMapping("/google")
    fun google(@Valid @RequestBody req: GoogleAuthRequest): ResponseEntity<ApiResponse<AuthResponse>> {
        return ResponseEntity.ok(ApiResponse.ok(authService.googleSignIn(req.idToken)))
    }

    @PostMapping("/apple")
    fun apple(@Valid @RequestBody req: AppleAuthRequest): ResponseEntity<ApiResponse<AuthResponse>> {
        return ResponseEntity.ok(
            ApiResponse.ok(
                authService.appleSignIn(req.identityToken, req.displayName, req.email)
            )
        )
    }

    @PostMapping("/refresh")
    fun refresh(@Valid @RequestBody req: RefreshTokenRequest): ResponseEntity<ApiResponse<AuthResponse>> {
        return ResponseEntity.ok(ApiResponse.ok(authService.refresh(req.refreshToken)))
    }

    @PostMapping("/logout")
    fun logout(@Valid @RequestBody req: LogoutRequest, auth: Authentication): ResponseEntity<Void> {
        val userId = UUID.fromString(auth.name)
        authService.logout(userId, req.refreshToken)
        return ResponseEntity.noContent().build()
    }
}

File: src/auth/auth.controller.ts
import { Body, Controller, HttpCode, HttpStatus, Post } from '@nestjs/common';
import { AuthService } from './auth.service';
import { GoogleAuthDto } from './dto/google-auth.dto';
import { AppleAuthDto } from './dto/apple-auth.dto';
import { RefreshTokenDto } from './dto/refresh-token.dto';
import { LogoutDto } from './dto/logout.dto';
import { AuthResponseDto } from './dto/auth-response.dto';

@Controller('v1/auth')
export class AuthController {
  constructor(private readonly authService: AuthService) {}

  @Post('google')
  @HttpCode(HttpStatus.OK)
  async googleSignIn(@Body() dto: GoogleAuthDto): Promise<AuthResponseDto> {
    return this.authService.googleSignIn(dto.idToken);
  }

  @Post('apple')
  @HttpCode(HttpStatus.OK)
  async appleSignIn(@Body() dto: AppleAuthDto): Promise<AuthResponseDto> {
    return this.authService.appleSignIn(
      dto.identityToken,
      dto.displayName,
      dto.email,
    );
  }

  @Post('refresh')
  @HttpCode(HttpStatus.OK)
  async refresh(@Body() dto: RefreshTokenDto): Promise<AuthResponseDto> {
    return this.authService.refresh(dto.refreshToken);
  }

  @Post('logout')
  @HttpCode(HttpStatus.OK)
  async logout(@Body() dto: LogoutDto): Promise<void> {
    // userId is not required for logout — we match by refreshToken only
    await this.authService.logout('', dto.refreshToken);
  }
}
